packages/router/src/lib/operations/metaTx.ts

```diff
@@ -0,0 +1,267 @@
+import {
+  ajv,
+  createLoggingContext,
+  MetaTxFulfillPayload,
+  MetaTxPayload,
+  MetaTxRouterContractCancelPayload,
+  MetaTxRouterContractFulfillPayload,
+  MetaTxRouterContractPreparePayload,
+  MetaTxType,
+  MetaTxTypes,
+  RequestContext,
+} from "@connext/nxtp-utils";
+import { BigNumber, constants, providers } from "ethers/lib/ethers";
+
+import { getContext } from "../../router";
+import { MetaTxInputSchema } from "../entities";
+import {
+  InvalidMetaTxType,
+  NoChainConfig,
+  NotAllowedFulfillRelay,
+  NotEnoughRelayerFee,
+  ParamsInvalid,
+} from "../errors";
+import { calculateGasFee, calculateGasFeeInReceivingTokenForFulfill } from "../helpers/shared";
+
+export const sendMetaTx = async <T extends MetaTxType>(
+  input: MetaTxPayload<T>,
+  _requestContext: RequestContext<string>,
+): Promise<providers.TransactionReceipt | undefined> => {
+  const { requestContext, methodContext } = createLoggingContext(sendMetaTx.name, _requestContext);
+
+  const { logger, contractWriter, config, chainData, txService } = getContext();
+  logger.debug("Method start", requestContext, methodContext, { input });
+
+  // Validate Input schema
+  const validateInput = ajv.compile(MetaTxInputSchema);
+  const validInput = validateInput(input);
+  if (!validInput) {
+    const msg = validateInput.errors?.map((err: any) => `${err.instancePath} - ${err.message}`).join(",");
+    throw new ParamsInvalid({
+      methodContext,
+      requestContext,
+      paramsError: msg,
+    });
+  }
+
+  const { chainId, data, type, to } = input;
+
+  let receipt;
+
+  if (!config.chainConfig[chainId]) {
+    throw new NoChainConfig(chainId, { methodContext, requestContext, input });
+  }
+
+  const relayerFeeLowerBound = config.chainConfig[chainId].relayerFeeThreshold;
+  if (!config.allowRelay && !config.chainConfig[chainId].allowRelay) {
+    // TODO: should we throwing error if relayer is not allowed?
+    // router shouldn't listen to metaTx request in first place if it's not allowed to relay
+    throw new NotAllowedFulfillRelay(chainId, {
+      methodContext,
+      requestContext,
+      data,
+      input,
+    });
+  }
+
+  // receiver fulfill, done directly on contract using user's sig broadcast
+  if (type === MetaTxTypes.Fulfill) {
+    const { txData, signature, relayerFee, callData } = data as MetaTxFulfillPayload;
+    // Send to tx service
+    logger.info("Sending fulfill tx", requestContext, methodContext, { signature });
+
+    let outputDecimals = chainData.get(txData.receivingChainId.toString())?.assetId[txData.receivingAssetId]?.decimals;
+    if (!outputDecimals) {
+      outputDecimals = await txService.getDecimalsForAsset(txData.receivingChainId, txData.receivingAssetId);
+    }
```

Comment:
really should just make a helper for this, we use this same logic everywhere

Reply:
Can this go into the txService?

Reply:
chaindata should get integrated into txservice imo, then we can combine that with the caching so it's just one call here `getDecimalsForAsset`